File: graph/src/main/groovy/net/twisterrob/gradle/graph/vis/d3/javafx/D3JavaFXTaskVisualizer.java
package net.twisterrob.gradle.graph.vis.d3.javafx;

import java.util.Map;

import javafx.event.EventHandler;
import javafx.stage.*;
import net.twisterrob.gradle.graph.tasks.*;
import net.twisterrob.gradle.graph.vis.TaskVisualizer;
import net.twisterrob.gradle.graph.vis.d3.javafx.Settings.WindowLocation;
import org.gradle.api.*;
import org.gradle.cache.PersistentCache;

/**
 * @see <a href="http://stackoverflow.com/a/20125944/253468">Idea from SO</a>
 */
public class D3JavaFXTaskVisualizer implements TaskVisualizer {
	private GraphWindow window;
	private final Settings settings;

	public D3JavaFXTaskVisualizer(PersistentCache cache) {
		settings = new Settings(cache);
	}

	@Override public void showUI(org.gradle.api.initialization.Settings project) {
		JavaFXApplication.startLaunch(settings.getSettings());
		window = JavaFXApplication.show(project);
		if (window != null) {
			settings.getSettings().applyTo(window.getStage());
			window.getStage().setOnHiding(new EventHandler<WindowEvent>() {
				@Override public void handle(WindowEvent event) {
					settings.setSettings(new WindowLocation((Window)event.getSource()));
					settings.close();
				}
			});
		}
	}

	@Override public void initModel(Map<Task, TaskData> graph) {
		if (window != null) {
			window.initModel(graph);
		}
	}

	@Override public void update(Task task, TaskResult result) {
		if (window != null) {
			window.update(task, result);
		}
	}

	@Override public void closeUI() {
		if (window == null) {
			settings.close();
		}
		window = null;
		JavaFXApplication.hide();
	}
}

File: graph/src/main/kotlin/net/twisterrob/gradle/graph/vis/d3/javafx/D3JavaFXTaskVisualizer.kt
package net.twisterrob.gradle.graph.vis.d3.javafx

import javafx.stage.Window
import net.twisterrob.gradle.graph.tasks.TaskData
import net.twisterrob.gradle.graph.tasks.TaskResult
import net.twisterrob.gradle.graph.vis.TaskVisualizer
import org.gradle.api.Task
import org.gradle.api.initialization.Settings
import org.gradle.cache.PersistentCache

/**
 * See [Idea from SO](http://stackoverflow.com/a/20125944/253468).
 */
class D3JavaFXTaskVisualizer(cache: PersistentCache) : TaskVisualizer {

	private var window: GraphWindow? = null
	private val options: Options

	init {
		options = Options(cache)
	}

	override fun showUI(settings: Settings) {
		JavaFXApplication.startLaunch(options.options)
		window = JavaFXApplication.show(settings)?.also { window ->
			options.options.applyTo(window.stage)
			window.stage.setOnHiding { event ->
				options.options = Options.WindowLocation(event.source as Window)
				options.close()
			}
		}
	}

	override fun initModel(graph: Map<Task, TaskData>) {
		window?.initModel(graph)
	}

	override fun update(task: Task, result: TaskResult) {
		window?.update(task, result)
	}

	override fun closeUI() {
		if (window == null) {
			options.close()
		}
		window = null
		JavaFXApplication.hide()
	}
}
